src/main/java/org/apache/commons/crypto/stream/CryptoOutputStream.java

```diff
@@ -298,30 +301,47 @@ public boolean isOpen() {
     @Override
     public int write(ByteBuffer src) throws IOException {
         checkStream();
-        final int len = src.remaining();
-        int remaining = len;
+
+        if (transferInProgress && outBuffer.hasRemaining()) {
+            output.write(outBuffer);
+            return 0;
+        } else {
+            outBuffer.clear();
+            transferInProgress = false;
+        }
+
+        int bytesWritten = 0;
+        int remaining = src.remaining();
+
         while (remaining > 0) {
-            final int space = inBuffer.remaining();
-            if (remaining < space) {
+            int written;
+            int space = inBuffer.remaining();
+
+            if(remaining < space) {
                 inBuffer.put(src);
-                remaining = 0;
+                bytesWritten += remaining;
+                break;
             } else {
-                // to void copy twice, we set the limit to copy directly
                 final int oldLimit = src.limit();
                 final int newLimit = src.position() + space;
                 src.limit(newLimit);
-
                 inBuffer.put(src);
-
-                // restore the old limit
                 src.limit(oldLimit);
-
                 remaining -= space;
-                encrypt();
+                bytesWritten += space;
+
+                encryptNonBlocking();
+                written = output.write(outBuffer);
+
+                // Underlying output maybe blocking, break out and return.
+                if (written < space) {
+                    transferInProgress = true;
```

Comment:
Do you really need this flag? Seems to me that checking whether `output` has data left should be enough?